### ISSUE TITLE
Problems with gRPC/Protobuf on Windows

### ISSUE THREAD

**Issue by AlgoryxMartinT** [dotnet/vscode-csharp#3313]
## Issue Description ##
I'm trying to create a project using Protobuf and gRPC, but Omnisharp keeps failing.

## Steps to Reproduce ##
The project I have issues with is a lot bigger, but I have tried to create a minimal example.

1. Create a new project with the following layout:
```
.
├── test.csproj
├── grpc
     └── test.proto
```
where `test.csproj` looks like this:
``` csproj
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFrameworks>netcoreapp3.0</TargetFrameworks>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Grpc" Version="2.24.0"/>
    <PackageReference Include="Grpc.Tools" Version="2.24.0"/>
    <Protobuf Include="grpc/test.proto" ProtoRoot="grpc"/>
  </ItemGroup>
</Project>
```
and `test.proto` like this (though I don't think this matters):
``` protobuf
syntax = "proto3";
package test;

message Test {
  int32 testy = 1;
}
```

2. Press "Restore" when the prompt in the lower right corner says there are unresolved dependencies.

3. Wait for a few seconds.

## Expected Behavior ##
The project is loaded with no errors.

## Actual Behavior ##
A prompt appears in the lower right corner saying
> Some projects have trouble loading. Please review the output for more details."

See the Logs section for more logs

### OmniSharp log ###
<details>

```
Starting OmniSharp server at 2019-10-03, 18:36:30
    Target: c:\Users\martint\test

OmniSharp server started.
    Path: C:\Users\martint\.vscode\extensions\ms-vscode.csharp-1.21.3\.omnisharp\1.34.5-beta.24\OmniSharp.exe
    PID: 3916

[info]: OmniSharp.Stdio.Host
        Starting OmniSharp on Windows 6.2.9200.0 (x64)
[info]: OmniSharp.Services.DotNetCliService
        DotNetPath set to dotnet
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Located 2 MSBuild instance(s)
            1: Visual Studio Professional 2017 15.9.28307.518 - "C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\15.0\Bin"
            2: StandAlone 16.0 - "C:\Users\martint\.vscode\extensions\ms-vscode.csharp-1.21.3\.omnisharp\1.34.5-beta.24\.msbuild\Current\Bin"
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        MSBUILD_EXE_PATH environment variable set to 'C:\Users\martint\.vscode\extensions\ms-vscode.csharp-1.21.3\.omnisharp\1.34.5-beta.24\.msbuild\Current\Bin\MSBuild.exe'
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Registered MSBuild instance: StandAlone 16.0 - "C:\Users\martint\.vscode\extensions\ms-vscode.csharp-1.21.3\.omnisharp\1.34.5-beta.24\.msbuild\Current\Bin"
            CscToolExe = csc.exe
            CscToolPath = C:\Users\martint\.vscode\extensions\ms-vscode.csharp-1.21.3\.omnisharp\1.34.5-beta.24\.msbuild\Current\Bin\Roslyn
            MSBuildExtensionsPath = C:\Users\martint\.vscode\extensions\ms-vscode.csharp-1.21.3\.omnisharp\1.34.5-beta.24\.msbuild
            MSBuildToolsPath = C:\Users\martint\.vscode\extensions\ms-vscode.csharp-1.21.3\.omnisharp\1.34.5-beta.24\.msbuild\Current\Bin
[info]: OmniSharp.Cake.CakeProjectSystem
        Detecting Cake files in 'c:\Users\martint\test'.
[info]: OmniSharp.Cake.CakeProjectSystem
        Could not find any Cake files
[info]: OmniSharp.MSBuild.ProjectSystem
        No solution files found in 'c:\Users\martint\test'
[info]: OmniSharp.MSBuild.ProjectManager
        Queue project update for 'c:\Users\martint\test\test.csproj'
[info]: OmniSharp.Script.ScriptProjectSystem
        Detecting CSX files in 'c:\Users\martint\test'.
[info]: OmniSharp.Script.ScriptProjectSystem
        Could not find any CSX files
[info]: OmniSharp.WorkspaceInitializer
        Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.CSharpFormattingWorkspaceOptionsProvider, Order: 0
[info]: OmniSharp.MSBuild.ProjectManager
        Loading project: c:\Users\martint\test\test.csproj
[info]: OmniSharp.WorkspaceInitializer
        Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.RenameWorkspaceOptionsProvider, Order: 100
[info]: OmniSharp.WorkspaceInitializer
        Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.EditorConfigWorkspaceOptionsProvider, Order: 200
[info]: OmniSharp.WorkspaceInitializer
        Configuration finished.
[info]: OmniSharp.Stdio.Host
        Omnisharp server running using Stdio at location 'c:\Users\martint\test' on host 21208.
[info]: OmniSharp.MSBuild.ProjectManager
        Successfully loaded project file 'c:\Users\martint\test\test.csproj'.
[info]: OmniSharp.MSBuild.ProjectManager
        Adding project 'c:\Users\martint\test\test.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Update project: test
[info]: OmniSharp.MSBuild.ProjectManager
        Queue project update for 'c:\Users\martint\test\test.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Queue project update for 'c:\Users\martint\test\test.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Queue project update for 'c:\Users\martint\test\test.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Queue project update for 'c:\Users\martint\test\test.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Loading project: c:\Users\martint\test\test.csproj
[fail]: OmniSharp.MSBuild.ProjectLoader
        The specified task executable location "\windows_x64\protoc.exe" is invalid.
[warn]: OmniSharp.MSBuild.ProjectManager
        Failed to load project file 'c:\Users\martint\test\test.csproj'.
c:\Users\martint\test\test.csproj
C:\Users\martint\.nuget\packages\grpc.tools\2.24.0\build\_protobuf\Google.Protobuf.Tools.targets(264,5): Error: The specified task executable location "\windows_x64\protoc.exe" is invalid.

[info]: OmniSharp.MSBuild.ProjectManager
        Update project: test
```

</details>

### C# log ###
<details>

```
Getting latest OmniSharp version information
Downloading package 'Latest OmniSharp Version Information' (1 KB).................... Done!
```

</details>

## Environment information ##

**VSCode version**: 1.38.1
**C# Extension**: 1.21.3

<details><summary>Dotnet Information</summary>
.NET Core SDK (reflecting any global.json):
 Version:   3.0.100
 Commit:    04339c3a26

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18362
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.0.100\

Host (useful for support):
  Version: 3.0.0
  Commit:  7d57652f33

.NET Core SDKs installed:
  2.1.2 [C:\Program Files\dotnet\sdk]
  2.1.4 [C:\Program Files\dotnet\sdk]
  2.1.104 [C:\Program Files\dotnet\sdk]
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.500 [C:\Program Files\dotnet\sdk]
  2.1.505 [C:\Program Files\dotnet\sdk]
  2.2.109 [C:\Program Files\dotnet\sdk]
  3.0.100 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
</details>
<details><summary>Visual Studio Code Extensions</summary>

|Extension|Author|Version|
|---|---|---|
|autodocstring|njpwerner|0.3.0|
|cmake|twxs|0.0.17|
|cpptools|ms-vscode|0.25.1|
|csharp|ms-vscode|1.21.3|
|csharpfixformat|Leopotam|0.0.84|
|dotenv|mikestead|1.0.1|
|gitlens|eamodio|10.0.1|
|hide-gitignored|npxms|1.0.1|
|hopscotch|idleberg|0.2.2|
|Lua|keyring|0.0.9|
|markdown-all-in-one|yzhang|2.4.2|
|markdown-pdf|yzane|1.3.1|
|material-icon-theme|PKief|3.9.0|
|preview-vscode|searKing|2.0.3|
|python|ms-python|2019.9.34911|
|rainbow-csv|mechatroner|1.3.1|
|restructuredtext|lextudio|114.0.0|
|scope-info|siegebell|0.2.0|
|svg|jock|1.0.2|
|swig-tpl|zhangciwu|0.0.6|
|Theme-peel|gerane|0.0.2|
|trailing-spaces|shardulm94|0.3.1|
|vscode-lua|gccfeli|0.1.2|
|vscode-ros|ms-iot|0.6.0|
|xml|DotJoshJohnson|2.5.0|;

</details>

**Comment by martielo** [comment]
Do you have Visual Studio 2019 installed in your machine?
In case you don't, try installing Visual Studio 2019 build tools.

You can find it here: https://visualstudio.microsoft.com//downloads/

**Comment by martielo** [comment]
Do you have Visual Studio 2019 installed on your machine?
In case you don't, try installing Visual Studio 2019 build tools.

You can find it here: https://visualstudio.microsoft.com//downloads/

**Comment by AlgoryxMartinT** [comment]
I tried installing VS2019, including the build tools, but the issue remains. This is my new OmniSharp log:

<Details>

```
Starting OmniSharp server at 2019-10-04, 09:22:04
    Target: c:\Users\martint\test

OmniSharp server started.
    Path: C:\Users\martint\.vscode\extensions\ms-vscode.csharp-1.21.4\.omnisharp\1.34.5-beta.24\OmniSharp.exe
    PID: 11968

[info]: OmniSharp.Stdio.Host
        Starting OmniSharp on Windows 6.2.9200.0 (x64)
[info]: OmniSharp.Services.DotNetCliService
        DotNetPath set to dotnet
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Located 3 MSBuild instance(s)
            1: Visual Studio Professional 2017 15.9.28307.518 - "C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\15.0\Bin"
            2: Visual Studio Professional 2019 16.3.29326.143 - "C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Current\Bin"
            3: StandAlone 16.0 - "C:\Users\martint\.vscode\extensions\ms-vscode.csharp-1.21.4\.omnisharp\1.34.5-beta.24\.msbuild\Current\Bin"
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Registered MSBuild instance: Visual Studio Professional 2019 16.3.29326.143 - "C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Current\Bin"
[info]: OmniSharp.Cake.CakeProjectSystem
        Detecting Cake files in 'c:\Users\martint\test'.
[info]: OmniSharp.Cake.CakeProjectSystem
        Could not find any Cake files
[info]: OmniSharp.MSBuild.ProjectSystem
        No solution files found in 'c:\Users\martint\test'
[info]: OmniSharp.MSBuild.ProjectManager
        Queue project update for 'c:\Users\martint\test\test.csproj'
[info]: OmniSharp.Script.ScriptProjectSystem
        Detecting CSX files in 'c:\Users\martint\test'.
[info]: OmniSharp.Script.ScriptProjectSystem
        Could not find any CSX files
[info]: OmniSharp.WorkspaceInitializer
        Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.CSharpFormattingWorkspaceOptionsProvider, Order: 0
[info]: OmniSharp.MSBuild.ProjectManager
        Loading project: c:\Users\martint\test\test.csproj
[info]: OmniSharp.WorkspaceInitializer
        Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.RenameWorkspaceOptionsProvider, Order: 100
[info]: OmniSharp.WorkspaceInitializer
        Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.EditorConfigWorkspaceOptionsProvider, Order: 200
[info]: OmniSharp.WorkspaceInitializer
        Configuration finished.
[info]: OmniSharp.Stdio.Host
        Omnisharp server running using Stdio at location 'c:\Users\martint\test' on host 4340.
[info]: OmniSharp.MSBuild.ProjectManager
        Successfully loaded project file 'c:\Users\martint\test\test.csproj'.
[info]: OmniSharp.MSBuild.ProjectManager
        Adding project 'c:\Users\martint\test\test.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Update project: test
[info]: OmniSharp.MSBuild.ProjectManager
        Queue project update for 'c:\Users\martint\test\test.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Queue project update for 'c:\Users\martint\test\test.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Queue project update for 'c:\Users\martint\test\test.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Queue project update for 'c:\Users\martint\test\test.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Loading project: c:\Users\martint\test\test.csproj
[fail]: OmniSharp.MSBuild.ProjectLoader
        The specified task executable location "\windows_x64\protoc.exe" is invalid.
[warn]: OmniSharp.MSBuild.ProjectManager
        Failed to load project file 'c:\Users\martint\test\test.csproj'.
c:\Users\martint\test\test.csproj
C:\Users\martint\.nuget\packages\grpc.tools\2.24.0\build\_protobuf\Google.Protobuf.Tools.targets(264,5): Error: The specified task executable location "\windows_x64\protoc.exe" is invalid.

[info]: OmniSharp.MSBuild.ProjectManager
        Update project: test
```

</Details>

The C# log is identical to the one before.

**Comment by rchande** [comment]
I think the line
> C:\Users\martint\.nuget\packages\grpc.tools\2.24.0\build\_protobuf\Google.Protobuf.Tools.targets(264,5): Error: The specified task executable location "\windows_x64\protoc.exe" is invalid.

is the relevant one. Looks like the protobuf targets are trying to concatenate a path with a msbuild variable that's not populated. Can you look at that file and share that line?

**Comment by AlgoryxMartinT** [comment]
These are the lines which start with 264:
``` csproj
    <ProtoCompile Condition=" '@(_Protobuf_OutOfDateProto)' != '' "
      ToolExe="$(Protobuf_ProtocFullPath)"
      Generator="$(Protobuf_Generator)"
      Protobuf="%(_Protobuf_OutOfDateProto.Source)"
      ProtoPath="%(_Protobuf_OutOfDateProto.AdditionalImportDirs);$(Protobuf_StandardImportsPath);%(_Protobuf_OutOfDateProto.ProtoRoot)"
      ProtoDepDir="$(Protobuf_DepFilesPath)"
      OutputDir="%(_Protobuf_OutOfDateProto.OutputDir)"
      OutputOptions="%(_Protobuf_OutOfDateProto._OutputOptions)"
      GrpcPluginExe="%(_Protobuf_OutOfDateProto.GrpcPluginExe)"
      GrpcOutputDir="%(_Protobuf_OutOfDateProto.GrpcOutputDir)"
      GrpcOutputOptions="%(_Protobuf_OutOfDateProto._GrpcOutputOptions)"
    >
      <Output TaskParameter="GeneratedFiles" ItemName="_Protobuf_GeneratedFiles"/>
    </ProtoCompile>
```
I think what the error message refers to is the `ToolExe`, which is defined like this on line 69 and forward:
``` csproj
    <PropertyGroup>
      <!-- First try environment variable. -->
      <Protobuf_ToolsOs>$(PROTOBUF_TOOLS_OS)</Protobuf_ToolsOs>
      <Protobuf_ToolsCpu>$(PROTOBUF_TOOLS_CPU)</Protobuf_ToolsCpu>
      <Protobuf_ProtocFullPath>$(PROTOBUF_PROTOC)</Protobuf_ProtocFullPath>

      <!-- Next try OS and CPU resolved by ProtoToolsPlatform. -->
      <Protobuf_ToolsOs Condition=" '$(Protobuf_ToolsOs)' == '' ">$(_Protobuf_ToolsOs)</Protobuf_ToolsOs>
      <Protobuf_ToolsCpu Condition=" '$(Protobuf_ToolsCpu)' == '' ">$(_Protobuf_ToolsCpu)</Protobuf_ToolsCpu>
      <Protobuf_ProtocFullPath Condition=" '$(Protobuf_ProtocFullPath)' == '' and '$(Protobuf_ToolsOs)' == 'windows' "
           >$(Protobuf_PackagedToolsPath)\$(Protobuf_ToolsOs)_$(Protobuf_ToolsCpu)\protoc.exe</Protobuf_ProtocFullPath>
      <Protobuf_ProtocFullPath Condition=" '$(Protobuf_ProtocFullPath)' == '' "
           >$(Protobuf_PackagedToolsPath)/$(Protobuf_ToolsOs)_$(Protobuf_ToolsCpu)/protoc</Protobuf_ProtocFullPath>
    </PropertyGroup>
```

So it seems like the problem is that the `Protobuf_PackagedToolsPath` gets set to an empty string. It is defined in the corresponding props file: `Google.Protobuf.Tools.props` in the same folder as the targets file. Line 11 looks like this:

``` csproj
    <Protobuf_PackagedToolsPath>$( [System.IO.Path]::GetFullPath($(MSBuildThisFileDirectory)../../tools) )</Protobuf_PackagedToolsPath>
```
Not sure how to continue debugging here. I have very little experience with MSBuild files.

I'm attaching the two files for reference (with .txt appended)

[Google.Protobuf.Tools.targets.txt](https://github.com/OmniSharp/omnisharp-vscode/files/3702215/Google.Protobuf.Tools.targets.txt)
[Google.Protobuf.Tools.props.txt](https://github.com/OmniSharp/omnisharp-vscode/files/3702216/Google.Protobuf.Tools.props.txt)

**Comment by koliyo** [comment]
Still having problems with this, but I have found a workaround that seems to help
```xml
<Protobuf Condition="! $(MSBuildExtensionsPath.Contains('omnisharp'))" ...>
```

**Comment by AlgoryxMartinT** [comment]
After some testing we (me and @koliyo) have found that the `Google.Protobuf.Tools.targets` file seems to be parsed before `Google.Protobuf.Tools.props`, and that therefore the properties defined in `.props` is never used by `.targets`.

The way we found it is by adding an `<Error Message="Error from .targets"/>` tag in the targets file under the `Protobuf_ResolvePlatform` target, and some garbage (e.g. `fdjkfsjsfljksdf`) somewhere in the props file. OmniSharp then shows the targets file error before the props file error.


The workaround suggested by @koliyo only works when using the StandAlone MSBuild instance. When Visual Studio is installed that one will be used instead, and MSBuildExtensionsPath will not contain the string "omnisharp". A more robust (maybe) workaround we have found is to add these lines to `omnisharp.json`:

```json
  "msbuild": {
    "Configuration": "omnisharp"
  }
```

and then this condition to the `csproj` file:

```xml
<Protobuf Condition="'$(Configuration)' != 'omnisharp'" ...>
```
